### ISSUE TITLE
Connection fail when Cassandra TLS certificate only contains dnsName without IP

### ISSUE THREAD

**Issue by leerin-ruby** [gocql/gocql#1611]
Please answer these questions before submitting your issue. Thanks!

### What version of Cassandra are you using?
Not related. Go only.

### What version of Gocql are you using?
v1.0.0

### What version of Go are you using?
go version go1.17.5 linux/amd64

### What did you do?
I attempt to connect to a cluster using SSL/TLS with EnableHostVerification set to true.

I connect to the cluster using FQDN. This FQDN is referenced in the dnsName field of the certificate.

### What did you expect to see?
Connect to the cluster successfully

### What did you see instead?
gocql: unable to create session: control: unable to connect to initial hosts: x509: cannot validate certificate for 192.168.3.10 because it doesn't contain any IP SANs

Dig into the code, first I found following:
session.go:
`	for _, host := range hostMap {
		host := s.ring.addOrUpdate(host)`
ring.go:
`func (r *ring) addOrUpdate(host *HostInfo) *HostInfo {
	if existingHost, ok := r.addHostIfMissing(host); ok {
		existingHost.update(host)
		host = existingHost
	}
	return host
}`
host_source.go:
`func (h *HostInfo) update(from *HostInfo) {
	if h == from {
		return
	}

	h.mu.Lock()
	defer h.mu.Unlock()

	from.mu.RLock()
	defer from.mu.RUnlock()

	// autogenerated do not update
	if h.peer == nil {
		h.peer = from.peer
	}
	if h.broadcastAddress == nil {
		h.broadcastAddress = from.broadcastAddress
	}
	if h.listenAddress == nil {
		h.listenAddress = from.listenAddress
	}
	if h.rpcAddress == nil {
		h.rpcAddress = from.rpcAddress
	}
	if h.preferredIP == nil {
		h.preferredIP = from.preferredIP
	}
	if h.connectAddress == nil {
		h.connectAddress = from.connectAddress
	}
	if h.port == 0 {
		h.port = from.port
	}
	if h.dataCenter == "" {
		h.dataCenter = from.dataCenter
	}
	if h.rack == "" {
		h.rack = from.rack
	}
	if h.hostId == "" {
		h.hostId = from.hostId
	}
	if h.workload == "" {
		h.workload = from.workload
	}
	if h.dseVersion == "" {
		h.dseVersion = from.dseVersion
	}
	if h.partitioner == "" {
		h.partitioner = from.partitioner
	}
	if h.clusterName == "" {
		h.clusterName = from.clusterName
	}
	if h.version == (cassVersion{}) {
		h.version = from.version
	}
	if h.tokens == nil {
		h.tokens = from.tokens
	}
}`

When create session, even the passed host structure contains the hostname, when run existingHost.update() in ring.go, the hostname isn't saved in the existingHost. This caused it still use the IP not hostname when connecting Cassandra with TLS and failed with certificate validation as no IP SAN is included in the certificate.

I don't know if it is a design choice or a bug, but the certificate may only contains dnsName. How to deal with this scenario? 
---

If you are having connectivity related issues please share the following additional information

### Describe your Cassandra cluster
please provide the following information

- output of `nodetool status`
- output of `SELECT peer, rpc_address FROM system.peers`
- rebuild your application with the `gocql_debug` tag and post the output

**Comment by leerin-ruby** [comment]
@martin-sucha thank you for the response. let's wait for the future enhancement, either the user-deifned function or the dialer interface.

**Comment by jameshartig** [comment]
@leerin-ruby the hostname you're passing initially, is that hostname present in all of the nodes' certificates? If so, you can modify the `clusterCfg.SslOpts` and set `ServerName` on the embedded `*tls.Config` to hardcode the expected hostname for all of the nodes.

**Comment by leerin-ruby** [comment]
@jameshartig, yes, the hostname is in all nodes' certificates(actually, I am using containerized cassandra and application), and the serverName is set in the clusterCfg.SslOpts otherwise the verification will fail more earlier.

**Comment by martin-sucha** [comment]
Hmm, I wonder why it fails with the `TLSConfig.ServerName` is set. If TLSConfig.ServerName is not empty, gocql shouldn't overwrite it: https://github.com/gocql/gocql/blob/4d42aa3a5f690598a34453ba94f5a379c83f5c94/conn.go#L264

In any case, once https://github.com/gocql/gocql/pull/1629 is merged, it should allow you to customize the setup of the TLS session.

**Comment by eevans** [comment]
With #1629 merged, should this issue be considered resolved?

**Comment by leerin-ruby** [comment]
> With #1629 merged, should this issue be considered resolved?

I think so. You can resolve.

**Comment by OleksiienkoMykyta** [comment]
It looks like this has been resolved since HostDialer interface is introduced, and as @leerin-ruby mentioned, this issue can be closed as resolved.
@joao-r-reis, could you close it, please?
